src/libbson/tests/test-bson-vector.c

```diff
@@ -0,0 +1,1400 @@
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+#include <bson/bson.h>
+
+#include "bsonutil/bson-parser.h"
+#include "TestSuite.h"
+#include "test-conveniences.h"
+#include "json-test.h"
+
+#ifdef _MSC_VER
+#define SSCANF sscanf_s
+#else
+#define SSCANF sscanf
+#endif
+
+struct view_abi_reference_type {
+   void *data;
+   uint32_t data_len;
+   uint8_t header_0;
+   uint8_t header_1;
+};
+
+/* ABI: Make sure vector views have the expected size. */
+#define EXPECTED_VECTOR_VIEW_SIZE (sizeof (struct view_abi_reference_type))
+BSON_STATIC_ASSERT2 (sizeof_bson_vector_int8_const_view_t,
+                     sizeof (bson_vector_int8_const_view_t) == EXPECTED_VECTOR_VIEW_SIZE);
+BSON_STATIC_ASSERT2 (sizeof_bson_vector_int8_view_t, sizeof (bson_vector_int8_view_t) == EXPECTED_VECTOR_VIEW_SIZE);
+BSON_STATIC_ASSERT2 (sizeof_bson_vector_float32_const_view_t,
+                     sizeof (bson_vector_float32_const_view_t) == EXPECTED_VECTOR_VIEW_SIZE);
+BSON_STATIC_ASSERT2 (sizeof_bson_vector_float32_view_t,
+                     sizeof (bson_vector_float32_view_t) == EXPECTED_VECTOR_VIEW_SIZE);
+BSON_STATIC_ASSERT2 (sizeof_bson_vector_packed_bits_const_view_t,
+                     sizeof (bson_vector_packed_bits_const_view_t) == EXPECTED_VECTOR_VIEW_SIZE);
+BSON_STATIC_ASSERT2 (sizeof_bson_vector_packed_bits_view_t,
+                     sizeof (bson_vector_packed_bits_view_t) == EXPECTED_VECTOR_VIEW_SIZE);
+#undef EXPECTED_VECTOR_VIEW_SIZE
+
+typedef struct vector_json_test_case_t {
+   char *scenario_description, *scenario_test_key;
+   char *test_description, *test_dtype_hex_str, *test_dtype_alias_str, *test_canonical_bson_str;
+   bson_t *test_vector_array;
+   int64_t *test_padding;
+   bool *test_valid;
+} vector_json_test_case_t;
+
+static void
+translate_json_test_vector (bson_t *array_in, bson_t *array_out)
+{
+   // Patches JSON tests until DRIVERS-3095 is resolved:
+   // - convert "inf" to a double +Infinity
+   // - convert "-inf" to a double -Infinity
+
+   bson_iter_t iter_in;
+   BSON_ASSERT (bson_iter_init (&iter_in, array_in));
+   bson_array_builder_t *builder = bson_array_builder_new ();
+   while (bson_iter_next (&iter_in)) {
+      if (BSON_ITER_HOLDS_UTF8 (&iter_in) && 0 == strcmp ("inf", bson_iter_utf8 (&iter_in, NULL))) {
+         BSON_ASSERT (bson_array_builder_append_double (builder, INFINITY));
+      } else if (BSON_ITER_HOLDS_UTF8 (&iter_in) && 0 == strcmp ("-inf", bson_iter_utf8 (&iter_in, NULL))) {
+         BSON_ASSERT (bson_array_builder_append_double (builder, -INFINITY));
```

Comment:
```suggestion
      if (BSON_ITER_HOLDS_UTF8 (&iter_in) && 0 == strcmp ("inf", bson_iter_utf8 (&iter_in, NULL))) {
         BSON_ASSERT (bson_array_builder_append_double (builder, 1.0 / 0.0));
      } else if (BSON_ITER_HOLDS_UTF8 (&iter_in) && 0 == strcmp ("-inf", bson_iter_utf8 (&iter_in, NULL))) {
         BSON_ASSERT (bson_array_builder_append_double (builder, -1.0 / 0.0));
```
Consider using `double` arithmetic instead of `INFINITY` (`float`) to address `-Wdouble-promotion` warnings. Also applies to other instances of `INFINITY` in this file.

Reply:
Did not realize UBSan would prohibit these operations. That is unfortunate. 😢
I am fine with leaving it as-written and deferring the "`INFINITY` for `double`" problem to another PR.